### PR TITLE
Assigning Unique External IPs for Ingress Traffic

### PR DESCRIPTION
**Pull request by tpoitras**
@marun - there is still a lot of missing info here, I think. But I read through the trello card and the test cases and took my best first attempt. Now I need some help, please.

Please review the following, review for technical accuracy, and also look for questions left in code blocks and as **NEEDINFO**:

http://file.bne.redhat.com/tpoitras/2016/exposeport/openshift-enterprise/expose-port-BZ1368874/admin_guide/tcp_ingress_external_ports.html

^ I've enabled commenting directly in that HTML file, just hover over the text you want to comment on and a little comment button will appear. Or, just leave comments here in the source of this PR.

Additionally, does this apply to OpenShift Dedicated 3.3? or just OCP?

### CODE REVIEW

File: _build_cfg.yml [hunk 1]
```diff
@@ -445,6 +445,8 @@ Topics:
     Distros: openshift-origin,openshift-enterprise
   - Name: Limit Pod Access with Egress Firewall
     File: limit_pod_access_egress
+  - Name: Enabling Ingress Traffic by Exposing External Ports
```
Comment (marun) [inline _build_cfg.yml]: Consider rewording this description.  The feature enables assignment of a unique ip address to a service that, if routed correctly, enables external traffic to reach that service's ports.  'External Ports' doesn't make sense to me.

File: admin_guide/tcp_ingress_external_ports.adoc [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+:icons:
+:experimental:
+:toc: macro
+:toc-title:
+
+toc::[]
+
+== Overview
+
+In {product-title} 3.2, you could use `*externalIP*` in the router service to control which external IP addresses expose the port. However, the OpenShift router could only listen on ports 80 and 443 for routing and mapping to backends (which have protocols). Previously, routes did not enable you to specify an external port. Port 80 runs in HTTP mode and 443 runs in TCP mode.
```
Comment (marun) [inline admin_guide/tcp_ingress_external_ports.adoc]: I'm not sure it makes sense to conflate externalIP with the router.  If a user wanted to expose an http/https service, they should use the router.  If they wanted to expose arbitrary ports, they should use an externalIP.  The feature is intended to make externalIP assignment more reliable, not to change anything about the router.

File: admin_guide/tcp_ingress_external_ports.adoc [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+However, you may want to expose external TCP traffic to the outside world.
+Starting in {product-title} 3.3, cluser administrators can expose an external
```
Comment (marun) [inline admin_guide/tcp_ingress_external_ports.adoc]: This was previously possible.  Kube has supported manual external ip assignment for some time:

http://kubernetes.io/docs/user-guide/services/#external-ips

The feature is to enable automatic and manual assignment of an address guaranteed to be assigned to at most one service.

File: admin_guide/tcp_ingress_external_ports.adoc [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+port, enabling ingress traffic (TCP/UDP) routing to a definable backend.
+
+Exposing external ports allows you to allocate a unique IP address per service,
+rather than trying to manage the port space of a limited number of shared IP
+addresses. This ensures that each service can simply direct traffic from its
+chosen ports, regardless of the ports selected by other services.
+
+There are benefits to using `*IngressIP*` instead of `*externalIP*`:
+
+- An ingress IP is intended to be automatically assigned, and can only be assigned to one service.
```
Comment (marun) [inline admin_guide/tcp_ingress_external_ports.adoc]: It is possible to request an ingress ip by specifying 'loadBalancerIP' in the same way described in the kube docs:

http://kubernetes.io/docs/user-guide/services/#type-loadbalancer

As with the kube, the specified ip is only a request.  If the ip is already allocated or out of range, another ip may be assigned to the service.

File: admin_guide/tcp_ingress_external_ports.adoc [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+Exposing external ports allows you to allocate a unique IP address per service,
+rather than trying to manage the port space of a limited number of shared IP
+addresses. This ensures that each service can simply direct traffic from its
+chosen ports, regardless of the ports selected by other services.
+
+There are benefits to using `*IngressIP*` instead of `*externalIP*`:
+
+- An ingress IP is intended to be automatically assigned, and can only be assigned to one service. 
+- An external IP must be manually assigned, and conflicting assignments cannot be detected. Unfortunately, multiple services could potentially request the same IP, and the conflict would go undetected.
```
Comment (marun) [inline admin_guide/tcp_ingress_external_ports.adoc]: It is possible that multiple services could specify the same externalIP and not conflict if the ports they were exposing were disjoint.  The potential for conflict is with ip:port combinations, not ip's by themselves.  The feature sidesteps the problem by ensuring that services do not share ingress ip's and therefore can expose any port without fear of conflicting with another service's ports.

File: admin_guide/tcp_ingress_external_ports.adoc [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+== Restrictions
+
+To use `*IngressIP*`, you can either:
+
+- Select an IP address with the existing `*ExternalIP*` ingress.
```
Comment (marun) [inline admin_guide/tcp_ingress_external_ports.adoc]: I'm not sure what this is trying to say.  Was your intent to highlight that it is possible for a service to specify the desired ingress ip address via the loadBalancerIP field (mentioned in a previous comment)?

File: admin_guide/tcp_ingress_external_ports.adoc [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
++
+Or,
++
+- Have an IP address assigned from a pool. In this case, {product-title} implements a bare-metal version of the LoadBalancer service type and assigns IP addresses to the services.
```
Comment (marun) [inline admin_guide/tcp_ingress_external_ports.adoc]: I don't think this is an 'or' situation.  On bare metal, with an ingress ip range configured, a service of type LoadBalancer will have an ingress ip allocated to it if one is available.  If a service specified loadBalancerIP and the address is in the ingress range and not already allocated to another service, the service will be assigned that address.

File: admin_guide/tcp_ingress_external_ports.adoc [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+Or,
++
+- Have an IP address assigned from a pool. In this case, {product-title} implements a bare-metal version of the LoadBalancer service type and assigns IP addresses to the services. 
++ 
+[CAUTION]
+====
+You must ensure that the IP address pool you assign terminates at a node in your cluster. You can use the existing `*oadm ipfailover*` to configure IP failover for the external IPs.
+====
+
+Potential port clashes are handled on a first-come, first-served basis. If you
```
Comment (marun) [inline admin_guide/tcp_ingress_external_ports.adoc]: This is true for manually configured externalIP's, but externalIP's assigned by the ingress controller are guaranteed to be assigned to at most one service.  This ensures that there is no potential for conflict - both service A and service B can expose mongo on it's standard port without fear of conflict because their respective ingress IP addresses will differ.

File: admin_guide/tcp_ingress_external_ports.adoc [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+request a port, it is only available if it has not yet been assigned. For
+example, `MongoDB service A` requests port 27017, and then `MongoDB service B`
+requests the same port; the first request gets the port, and the other receives
+a random port. This is similar to the route name allocation mechanism. 
+
+[NOTE]
+====
+You can only expose external ports in a sharded router environment, and only in
```
Comment (marun) [inline admin_guide/tcp_ingress_external_ports.adoc]: This feature has nothing to do with the router, so the first requirement is incorrect.

File: admin_guide/tcp_ingress_external_ports.adoc [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+bare-metal environments. Ingress IPs can only be assigned if they are not
```
Comment (marun) [inline admin_guide/tcp_ingress_external_ports.adoc]: they are -> the cluster is

File: admin_guide/tcp_ingress_external_ports.adoc (hunk 9 above)
Comment (marun) [inline admin_guide/tcp_ingress_external_ports.adoc]: Also, the feature doesn't 'expose external ports'.  It allows externalIPs to be assigned uniquely to services to eliminate the possibility of ip:port conflicts.

File: admin_guide/tcp_ingress_external_ports.adoc [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+running in the cloud. In cloud environments, LoadBalancer-type services
+configure cloud-specific load balancers.
+====
+
+== Configuring an Ingress IP
+
+To assign an ingress IP:
+
+*NEEDINFO*: Do they first need to set `ingressIPNetworkCIDR` in *_master-config.yaml_*, and restart the master service?
```
Comment (marun) [inline admin_guide/tcp_ingress_external_ports.adoc]: Yes

File: admin_guide/tcp_ingress_external_ports.adoc [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+
+. Update the status of a LoadBalancer service on your pod to include an ingress IP:
```
Comment (marun) [inline admin_guide/tcp_ingress_external_ports.adoc]: Making a service of type LoadBalancer is sufficient.  Previously, services on a non-cloud cluster were likely to be of type ClusterIP or NodePort because LoadBalancer was only useful in a cloud environment.  The feature reuses the LoadBalancer type to enable assignment of non-conflicting external ip's.

File: admin_guide/tcp_ingress_external_ports.adoc [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
++
+----
+# How do they do this?
+----
+. Get the ingress IP:
```
Comment (marun) [inline admin_guide/tcp_ingress_external_ports.adoc]: The ingress ip will appear in the status field as per the kube docs on LoadBalancer:

http://kubernetes.io/docs/user-guide/services/#type-loadbalancer

It will also appear as an externalIP as per the kube docs:

http://kubernetes.io/docs/user-guide/services/#external-ips

It may be sufficient to say 'check the service for an external ip'.  When a service of type LoadBalancer hasn't yet had an ingress ip assigned, the expected output of 'oc get myservice' would be:


    NAME         CLUSTER-IP      EXTERNAL-IP   PORT(S)   AGE
    myservice    172.30.74.106   <pending>     80/TCP    30s 

Once an ingress ip was assigned, it would be:

    NAME         CLUSTER-IP      EXTERNAL-IP   PORT(S)   AGE
    myservice    172.30.74.106   172.46.0.1    80/TCP    30s

File: admin_guide/tcp_ingress_external_ports.adoc [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
++
+----
+# oc get svc
+----
+. Add the IP to the list of external IPs. For example:
```
Comment (marun) [inline admin_guide/tcp_ingress_external_ports.adoc]: This would be performed separately by the cluster administrator along with choosing an appropriate value for IngressIPNetworkCIDR for the cluster.  I think it makes sense to separate cluster configuration from configuring a service to be assigned an ingress ip.

File: install_config/install/deploy_router.adoc [hunk 15]
```diff
@@ -697,7 +697,6 @@ router pod, you should see the unlabelled effects.
 $ oc scale dc/router --replicas=0 && oc scale dc/router --replicas=1
 ----
 
-
```
Comment (marun) [inline install_config/install/deploy_router.adoc]: Consider reverting this unrelated change.

**Comment by marun** [comment]
I have some preliminary feedback.  If this is too disjoint I'd be up for having a call though the scheduling might be challenging given our timezones (assuming you're down under).

File: admin_guide/tcp_ingress_external_ports.adoc (hunk 6 above)
Comment (tpoitras) [inline admin_guide/tcp_ingress_external_ports.adoc]: I got that from this comment on the Trello card: 

**Ben Bennett**
_@jorgemoralespou: Maru posted openshift/origin: Pull Request 9454 and based on the feedback_
_is rewriting it. At the moment the strategy is that you can either choose an IP address with_
_the existing ExternalIP ingress, or if you want an IP address to be assigned from a pool, then_ 
_the new feature will implement a bare-metal version of the LoadBalancer service type and assign_ _IP addresses to the services. You will need to ensure that the IP address pool you assign_ _terminates at a node on your cluster. You can use the existing 'oadm ipfailover' to configure ip_ _failover for the external IPs._

Apologies if it's wrong to include that here, but the Trello card had zero information on what needed to be documented so I just did my best with what was there to use as a starting point.

File: admin_guide/tcp_ingress_external_ports.adoc (hunk 9 above)
Comment (tpoitras) [inline admin_guide/tcp_ingress_external_ports.adoc]: My docs BZ is titled "Document the ability to expose an external port other than 80 and 443" so I've got that "expose port" thing stuck in my head. Thanks for the correction.

**Comment by marun** [comment]
@tpoitras re: ocp vs dedicated, the feature is only supported on a non-cloud platform (vm or bare-metal).

File: _build_cfg.yml (hunk 1 above)
Comment (adellape) [inline _build_cfg.yml]: "Enabling Ingress Traffic"
"Assigning Unique IPs for Ingress Traffic"
"Unique Service IPs for Ingress Traffic"

?

Comment (marun) [inline _build_cfg.yml]: "Unique External IPs for Ingress Traffic"?

Comment (tpoitras) [inline _build_cfg.yml]: "Assigning Unique External IPs for Ingress Traffic"?

File: admin_guide/tcp_ingress_external_ports.adoc (hunk 6 above)
Comment (tpoitras) [inline admin_guide/tcp_ingress_external_ports.adoc]: @marun based on the above quote from the Trello card, can you suggest an edit to the existing line that'd make it clearer what it's trying to say?

File: admin_guide/tcp_ingress_external_ports.adoc (hunk 14 above)
Comment (tpoitras) [inline admin_guide/tcp_ingress_external_ports.adoc]: I've separated those 2 things out into their own sections before and after this section

**Comment by tpoitras** [comment]
@marun I do believe I have incorporated all your feedback, please take another look here and let me know if you find anything lacking or incorrect. Also, I do still have a question in there from the previous commit.

File: admin_guide/tcp_ingress_external_ports.adoc [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+
+There is support for both automatic and manual assignment of IP addresses, each
+address is guaranteed to be assigned to a limit of not more than one service.
+This ensures that each service can simply direct traffic from its chosen ports,
+regardless of the ports selected by other services.
+
+[[unique-external-ips-ingress-traffic-restrictions]]
+== Restrictions
+
+To use `*IngressIP*`, you can:
```
Comment (marun) [inline admin_guide/tcp_ingress_external_ports.adoc]: The 2 statements that follow better describe 'External IP'. An 'ingress ip' is an external ip guaranteed to be uniquely assigned to a service.  A manually-assigned external ip has no such guarantee.

'To use an ExternalIP...' ?

File: admin_guide/tcp_ingress_external_ports.adoc [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+
+- Select an IP address with the existing `*ExternalIP*` ingress.
+- Have an IP address assigned from a pool. In this case, {product-title} implements a bare-metal version of the LoadBalancer service type and assigns IP addresses to the services.
```
Comment (marun) [inline admin_guide/tcp_ingress_external_ports.adoc]: Consider changing all references to 'bare-metal' with 'non-cloud' or something similar, since the ingress controller will work on a vm-based cluster (technically non-cloud).

File: admin_guide/tcp_ingress_external_ports.adoc [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+:icons:
+:experimental:
+:toc: macro
+:toc-title:
+
+toc::[]
+
+== Overview
+
+Cluster administrators can assign a unique IP address to a service that, if
```
Comment (marun) [inline admin_guide/tcp_ingress_external_ports.adoc]: unique -> unique external

File: admin_guide/tcp_ingress_external_ports.adoc [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+routed correctly, enables external traffic to reach the ports of that service.
+This external IP assignment allows you to expose an external port, enabling
```
Comment (marun) [inline admin_guide/tcp_ingress_external_ports.adoc]: This...backend. -> (remove)

File: admin_guide/tcp_ingress_external_ports.adoc [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+:experimental:
+:toc: macro
+:toc-title:
+
+toc::[]
+
+== Overview
+
+Cluster administrators can assign a unique IP address to a service that, if
+routed correctly, enables external traffic to reach the ports of that service.
```
Comment (marun) [inline admin_guide/tcp_ingress_external_ports.adoc]: the ports of that service -> that service's endpoints via any TCP and UDP port the service exposes. ?

**Review by marun — COMMENTED** [review]
Hopefully I've answered your questions, more comments/suggestions inline.  Anything with a '?' is entirely suggestion only and need be applied only at your discretion.  Thank you!

File: admin_guide/tcp_ingress_external_ports.adoc [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+
+In non-cloud clusters, `ingressIPNetworkCIDR` is set by default to `172.46.0.0/16`. You could use the default if your cluster environment is not already using this private range. However, if you want to use a different range, then before you assign an ingress IP you must set xref:../install_config/master_node_configuration.adoc#master-node-config-network-config[`ingressIPNetworkCIDR`]
+in the *_/etc/origin/master-config.yaml_* file, then restart the master service.
+
+[[unique-external-ips-ingress-traffic-configure-service]]
+== Configuring an Ingress IP for a Service
+
+To assign an ingress IP:
+
+. Create a LoadBalancer service on your pod.
```
Comment (tpoitras) [inline admin_guide/tcp_ingress_external_ports.adoc]: nah, we try to avoid the s-in-brackets thing.

**Comment by tpoitras** [comment]
Thank you, @marun :+1: 
I've made updates based on your suggestions, which you can preview here: http://file.bne.redhat.com/tpoitras/2016/exposeport/openshift-enterprise/expose-port-BZ1368874/admin_guide/tcp_ingress_external_ports.html

Please let me know if you find anything lacking or incorrect. Otherwise, if you give it the thumbs-up, I'll move it along to the next stage of review. Thanks! :octocat:

**Comment by marun** [comment]
@tpoitras My apologies, but I do have further comments.  Please update this PR so I can make them inline.

File: admin_guide/tcp_ingress_external_ports.adoc (hunk 21 above)
Comment (eparis) [inline admin_guide/tcp_ingress_external_ports.adoc]: Can we get and example service yaml? In particular we should show the `LoadBalancer` type.
```
apiVersion: v1
kind: Service
metadata:
  name: egress-1
spec:
  ports:
  - name: db
    port: 3306
  type: LoadBalancer
  selector:
    name: my-db-selector
```

File: admin_guide/tcp_ingress_external_ports.adoc [hunk 22]
```diff
@@ -0,0 +1,110 @@
...
+To use an `*ExternalIP*`, you can:
+
+- Select an IP address from the `*ExternalIPNetworkCIDRs*` range. 
+- Have an IP address assigned from a pool. In this case, {product-title} implements a non-cloud version of the LoadBalancer service type and assigns IP addresses to the services.
++
+[CAUTION]
+====
+You must ensure that the IP address pool you assign terminates at one or more
+nodes in your cluster. You can use the existing `*oadm ipfailover*` to ensure
+that the external IPs are highly available.
```
Comment (eparis) [inline admin_guide/tcp_ingress_external_ports.adoc]: Is there a link we could put to how to use `oadm ipfailover` from the router HA docs?

Comment (tpoitras) [inline admin_guide/tcp_ingress_external_ports.adoc]: Good idea @eparis 
I've made the update, thanks

File: admin_guide/tcp_ingress_external_ports.adoc (hunk 21 above)
Comment (tpoitras) [inline admin_guide/tcp_ingress_external_ports.adoc]: @eparis - thanks for that! What would be the command to use this yaml to create the service?

**Comment by tpoitras** [comment]
@marun 
RE: "Please update this PR", my changes were already here, but as a new commit, which may have been unclear, so I've squashed everything down into one commit now.

**Comment by tpoitras** [comment]
@eparis - thank you for the example yaml, would you be able to indicate the command required to use that yaml to create a loadbalancer service?

**Comment by eparis** [comment]
I believe it would just be `oc create -f $file_with_that.yaml`

**Comment by tpoitras** [comment]
Thank you, @eparis - I have updated again.

File: admin_guide/tcp_ingress_external_ports.adoc [hunk 23]
```diff
@@ -0,0 +1,132 @@
...
+first-come, first-served basis. If you request a port, it is only available if
+it has not yet been assigned for that IP address. For example:
+
+.Port clash example for manually-configured external IPs
+====
+Two services have been manually configured with the same external
+IP address of 172.7.7.7.
+
+`MongoDB service A` requests port 27017, and then
+`MongoDB service B` requests the same port; the first request gets the port, and
```
Comment (marun) [inline admin_guide/tcp_ingress_external_ports.adoc]: It is not like route name allocation.  If an external ip:port combination is already bound on a node by kube-proxy, subsequent attempts to bind that combination will fail.   An error will be logged, but a random port will not be assigned, and there will be no indication on the service that its ports are not being exposed.  This was the motivation behind automating allocation of unique external ip's.

File: admin_guide/tcp_ingress_external_ports.adoc [hunk 24]
```diff
@@ -0,0 +1,132 @@
...
+----
+
+[[unique-external-ips-ingress-traffic-routing-cidr]]
+== Routing the Ingress CIDR for Development or Testing
+
+Add a static route directing traffic for the ingress CIDR to a node in the
+cluster. For example:
+
+----
+# route add -net 172.31.0.0/16 gw 10.66.140.17 eth0
```
Comment (marun) [inline admin_guide/tcp_ingress_external_ports.adoc]: Consider changing the ingress cidr to the default, 172.46.0.0/16.

File: admin_guide/tcp_ingress_external_ports.adoc [hunk 25]
```diff
@@ -0,0 +1,132 @@
...
+in the *_/etc/origin/master-config.yaml_* file, then restart the master service.
+
+[[unique-external-ips-ingress-traffic-configure-service]]
+== Configuring an Ingress IP for a Service
+
+To assign an ingress IP:
+
+. Create a YAML file for a LoadBalancer service:
++
+.Sample LoadBalancer Configuration
```
Comment (marun) [inline admin_guide/tcp_ingress_external_ports.adoc]: Consider requesting that a specific ip be assigned by setting loadBalancerIP as per the kube docs:

http://kubernetes.io/docs/user-guide/services/#type-loadbalancer

File: admin_guide/tcp_ingress_external_ports.adoc [hunk 26]
```diff
@@ -0,0 +1,132 @@
...
+:toc: macro
+:toc-title:
+
+toc::[]
+
+== Overview
+
+Cluster administrators can assign a unique external IP address to a service. If
+routed correctly, external traffic can reach that service's endpoints via any
+TCP/UDP port the service exposes. This can be simpler than trying to manage the
```
Comment (marun) [inline admin_guide/tcp_ingress_external_ports.adoc]: trying -> having

File: admin_guide/tcp_ingress_external_ports.adoc [hunk 27]
```diff
@@ -0,0 +1,132 @@
...
+port space of a limited number of shared IP addresses, and manually assigning
```
Comment (marun) [inline admin_guide/tcp_ingress_external_ports.adoc]: ,and manually -> when manually

**Review by marun — COMMENTED** [review]
My apologies, I was confused by github's new review function.  I'm used to comments being hidden when a PR is updated but comments from a review persist.

File: admin_guide/tcp_ingress_external_ports.adoc (hunk 23 above)
Comment (tpoitras) [inline admin_guide/tcp_ingress_external_ports.adoc]: Thanks, I'll remove the comparison to route name allocation, as well as the part about the random port.

**Comment by tpoitras** [comment]
Thank you, @marun - changes made as per your notes

Moving along to docs QA review:
http://file.bne.redhat.com/tpoitras/2016/exposeport/openshift-enterprise/expose-port-BZ1368874/admin_guide/tcp_ingress_external_ports.html

File: admin_guide/tcp_ingress_external_ports.adoc [hunk 28]
```diff
@@ -0,0 +1,130 @@
...
+====
+. Create a LoadBalancer service on your pod:
++
+----
+$ oc create -f loadbalancer.yaml
+----
+. Check the service for an external IP. For example, for a service named `myservice`:
++
+----
+$ oc get myservice
```
Comment (zhaozhanqi) [inline admin_guide/tcp_ingress_external_ports.adoc]: oc get myservice  --> oc get svc myservice

File: admin_guide/tcp_ingress_external_ports.adoc [hunk 29]
```diff
@@ -0,0 +1,130 @@
...
+====
+Ingress IPs can only be assigned if the cluster is not running in the cloud. In
+cloud environments, LoadBalancer-type services configure cloud-specific load
+balancers.
+====
+
+[[unique-external-ips-ingress-traffic-configure-cluster]]
+== Configuring the Cluster to Use Unique External IPs
+
+In non-cloud clusters, `ingressIPNetworkCIDR` is set by default to `172.46.0.0/16`. You could use the default if your cluster environment is not already using this private range. However, if you want to use a different range, then before you assign an ingress IP you must set xref:../install_config/master_node_configuration.adoc#master-node-config-network-config[`ingressIPNetworkCIDR`]
```
Comment (zhaozhanqi) [inline admin_guide/tcp_ingress_external_ports.adoc]: the feature 'ingressIPNetworkCIDR' is off as default when installing openshift env using ansible for now. I reported one bug for trace this https://bugzilla.redhat.com/show_bug.cgi?id=1366125

File: admin_guide/tcp_ingress_external_ports.adoc [hunk 30]
```diff
@@ -0,0 +1,130 @@
...
+. Create a YAML file for a LoadBalancer service that requests a specific IP via the `*loadBalancerIP*` setting:
++
+.Sample LoadBalancer Configuration
+====
+----
+apiVersion: v1
+kind: Service
+metadata:
+  name: egress-1
+spec: ports:
```
Comment (zhaozhanqi) [inline admin_guide/tcp_ingress_external_ports.adoc]: the format is incorrect: spec:ports:  'ports' should be in newline. you can see:
apiVersion: v1
kind: Service
metadata:
  name: egress-1
spec:
  ports:
  - name: db
    port: 3306
  loadBalancerIP: 172.46.0.1
  type: LoadBalancer
  selector:
    name: my-db-selector

File: admin_guide/tcp_ingress_external_ports.adoc [hunk 31]
```diff
@@ -0,0 +1,130 @@
...
+When your LoadBalancer-type service has an external IP assigned, the output
+displays the IP:
++
+----
+NAME         CLUSTER-IP      EXTERNAL-IP   PORT(S)   AGE
+myservice    172.30.74.106   172.46.0.1    80/TCP    30s
+----
++
+However, if your LoadBalancer-type service does not have a properly-assigned
+ingress IP, the output shows it as pending:
```
Comment (zhaozhanqi) [inline admin_guide/tcp_ingress_external_ports.adoc]: since the default  'ingressIPNetworkCIDR' is 172.46.0.0/16.  so if not setting 'loadBalancerIP' in service yaml file. the 'EXTERNAL-IP' will be assigned one ip automatically from the range.

Comment (tpoitras) [inline admin_guide/tcp_ingress_external_ports.adoc]: ah okay cool, so I'll remove this part then

**Comment by tpoitras** [comment]
@zhaozhanqi 
I've made updates, please check again and let me know if this passes now

**Comment by vikram-redhat** [comment]
This has passed QE: https://bugzilla.redhat.com/show_bug.cgi?id=1368874#c8. Merging it.

File: admin_guide/tcp_ingress_external_ports.adoc [hunk 32]
```diff
@@ -0,0 +1,123 @@
...
+ports exposed by other services.
+
+[[unique-external-ips-ingress-traffic-restrictions]]
+== Restrictions
+
+To use an `*ExternalIP*`, you can:
+
+- Select an IP address from the `*ExternalIPNetworkCIDRs*` range.
+- Have an IP address assigned from a pool. In this case, {product-title} implements a non-cloud version of the LoadBalancer service type and assigns IP addresses to the services.
++
```
Comment (bfallonf) [inline admin_guide/tcp_ingress_external_ports.adoc]: I'm not sure what the above means. Is the `*ExternalIP*` a file somewhere?

**Comment by bfallonf** [comment]
@tpoitras Just one comment.
